### PR TITLE
videoio: use oneVPL load mechanism, encoder bitrate estimation

### PR DESCRIPTION
**Pull request by mshabunin**
* use oneVPL [MFXLoad](https://spec.oneapi.io/versions/latest/elements/oneVPL/source/API_ref/VPL_disp_api_func.html#_CPPv47MFXLoadv) function instead of deprecated _MFXInit_ 
* advanced encoding bitrate estimation with different formulas for MPEG2, H264 and H265
* minor modification to libva search process

- [x] Check on Linux with custom oneVPL build (https://github.com/mshabunin/onevpl-opencv-check).
- [x] Check on Windows with custom oneVPL build

<cut/>

Troubleshooting:

- (Linux) crash with following message might be caused by outdated kernel (observed with Linux 4.4.0)

    ```
    libva info: Found init function __vaDriverInit_1_12
    libva info: va_openDriver() returns 0
    opencv_test_videoio: /opt/src/msdk/_studio/mfx_lib/shared/src/mfx_mpeg2_encode_vaapi.cpp:1420: mfxStatus 
    MfxHwMpeg2Encode::VAAPIEncoder::Execute(MfxHwMpeg2Encode::ExecuteBuffers*, mfxU32, mfxU8*, mfxU32): Assertion 
    `0x00000000 == vaSts' failed.
    Process returned: -11
    ```

    need to upgrade Linux kernel

---

```
force_builders=Custom,Custom Win

Xbuildworker:Custom=linux-1
buildworker:Custom=linux-6
build_image:Custom=onevpl-2021.6.0
test_opencl:Custom=ON
test_modules:Custom=videoio


buildworker:Custom Win=windows-3
build_image:Custom Win=onevpl-2021.6.0
test_opencl:Custom Win=ON
test_modules:Custom Win=videoio
```

### CODE REVIEW

File: modules/videoio/src/cap_mfx_common.hpp [hunk 1]
```diff
@@ -176,10 +177,38 @@ inline void cleanup(T * &ptr)
 
 //==================================================================================================
 
+class MFXVideoSession_WRAP : public MFXVideoSession
```
Comment (alalek) [inline modules/videoio/src/cap_mfx_common.hpp:180]: I believe, we should not embed factory behavior into objects (through inheritance). Separate classes for that should be more suitable.

File: modules/videoio/src/cap_mfx_common.hpp [hunk 2]
```diff
@@ -176,10 +177,38 @@ inline void cleanup(T * &ptr)
...
+        MFXSetConfigFilterProperty(cfg, (const mfxU8*)"mfxImplDescription.Impl", impl);
+    }
+    ~MFXVideoSession_WRAP()
+    {
+        MFXUnload(loader);
+    }
+    mfxStatus Load()
+    {
+        return MFXCreateSession(loader, 0, &m_session);
+    }
```
[the remark this answers is not in the document]
Comment (mshabunin) [reply modules/videoio/src/cap_mfx_common.hpp:200]: `MFXInit` is used in MediaSDK to create a session: https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdk-man.md#MFXInit

I'll rename this method to `CreateSession`.

File: modules/videoio/src/cap_mfx_common.hpp (hunk 1 above)
Comment (mshabunin) [reply modules/videoio/src/cap_mfx_common.hpp:180]: This class extends existing MediaSDK/oneVPL thin C++ wrapper which stores a session and provides analogs of C-functions via methods (https://github.com/Intel-Media-SDK/MediaSDK/blob/master/api/include/mfxvideo%2B%2B.h). `MFXVideoSession` already had this behavior (`Init`/`Close`), I've only added a new method to replace `Init`.

I'll think about moving loader out of this class and creating a new one (or singleton).

File: modules/videoio/src/cap_mfx_common.hpp [hunk 3]
```diff
@@ -176,10 +177,38 @@ inline void cleanup(T * &ptr)
...
+        loader = MFXLoad();
+        mfxConfig cfg = MFXCreateConfig(loader);
+        mfxVariant impl;
+        impl.Type = MFX_VARIANT_TYPE_U32;
+        impl.Data.U32 = MFX_IMPL_TYPE_HARDWARE;
+        MFXSetConfigFilterProperty(cfg, (const mfxU8*)"mfxImplDescription.Impl", impl);
+    }
+    ~MFXVideoSession_WRAP()
+    {
+        MFXUnload(loader);
```
[the remark this answers is not in the document]
Comment (mshabunin) [reply modules/videoio/src/cap_mfx_common.hpp:195]: I think singleton is not necessary here, it depends on how heavy loader creation process is. For now it seems that creating and destroying loader each time does not add an overhead and it is appropriate to always load HW-accelerated implementation.

Later, if we decide to implement implementation enumeration capability, situation might change.

I'll investigate loading process a little bit more.

[the remark this answers is not in the document]
Comment (mshabunin) [reply modules/videoio/src/cap_mfx_common.hpp]: Looks like using singleton slightly reduces library loading overhead, so I've extracted loader creation as suggested.

File: modules/videoio/src/cap_mfx_common.hpp (hunk 1 above)
Comment (mshabunin) [reply modules/videoio/src/cap_mfx_common.hpp:186]: Loader creation have been extracted.

File: modules/videoio/src/cap_mfx_common.hpp [hunk 4]
```diff
@@ -18,6 +18,7 @@
 #  include <vpl/mfxvideo++.h>
```
[the remark this answers is not in the document]
Comment (mshabunin) [reply modules/videoio/src/cap_mfx_common.hpp:18]: I've added the check. Verified it with oneVPL 2021.1.1-66, 2021.2.2-212, 2021.4.0-328 from oneAPI and 2021.4.0 from [GPGPU release](https://dgpu-docs.intel.com/releases/releases-20210817.html). Support of 2021.1 has been successfully disabled. Additionally I've found that oneVPL has changed versioning scheme and 2021.4.0 is now "2.4", thus slightly more complex condition.

[the remark this answers is not in the document]
Comment (alalek) [reply modules/videoio/src/cap_mfx_common.hpp:18]: > I've found that oneVPL has changed versioning scheme and 2021.4.0 is now "2.4"

🤦‍♂️ 
Confirmed.

```
VPL_CONFIG=C:/Program Files (x86)/Intel/oneAPI/vpl/2021.4.0/lib/cmake/vpl/VPLConfig.cmake
...
VPL_VERSION=2.4
```

(enabled "Custom Win" builder, vpl-inspect dump is available on "init" step)

File: modules/videoio/src/cap_mfx_writer.cpp [hunk 5]
```diff
@@ -11,6 +11,31 @@
...
+        bitrate = (mp * 43) * fps + 360;
+    }
+    else if (codecId == MFX_CODEC_AVC)
+    {
+        bitrate = (mp * 140 + 19) * pow(fps, 0.60f);
+    }
+    else if (codecId == MFX_CODEC_HEVC)
+    {
+        bitrate = (mp * 63 + 41) * pow(fps, 0.60f);
+    }
```
Comment (alalek) [inline modules/videoio/src/cap_mfx_writer.cpp:29]: Do you know what is wrong with .h265 tests on "Custom Win" builder?

**Review by alalek — COMMENTED** [review]
It make sense to eliminate/suppress deprecation build warnings.

File: modules/videoio/src/cap_mfx_writer.cpp (hunk 5 above)
Comment (mshabunin) [reply modules/videoio/src/cap_mfx_writer.cpp:29]: I'm not sure what the problem is. In my tests on laptop with UHD 620 (27.20.100.9466) and oneVPL 2021.4.0 from oneAPI all MFX tests are green. Perhaps some combinations of HW / Driver / oneVPL are not working very well.

Comment (mshabunin) [reply modules/videoio/src/cap_mfx_writer.cpp]: I've increased bitrate slightly for HEVC. Let's see, maybe it will help.

Comment (alalek) [reply modules/videoio/src/cap_mfx_writer.cpp]: Looks like there is mess in frames order.
Attached captured frames and encoded videos of failed test cases: [pr20614.zip](https://github.com/opencv/opencv/files/7271777/pr20614.zip)

BTW, encoded videos are broken (check `ffmpeg -i 1.265 -r 1/1 frame_%05d.png`)

Comment (mshabunin) [reply modules/videoio/src/cap_mfx_writer.cpp]: I don't see suspicious messages in the ffmpeg output:
```
$ ffmpeg -i 1.265  -r 1/1 frame_%05d.png
ffmpeg version 4.3.2-0+deb11u1ubuntu1 Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 10 (Ubuntu 10.2.1-20ubuntu1)
  configuration: --prefix=/usr --extra-version=0+deb11u1ubuntu1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-avresample --disable-filter=resample --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-pocketsphinx --enable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-nvenc --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
  libpostproc    55.  7.100 / 55.  7.100
Input #0, hevc, from '1.265':
  Duration: N/A, bitrate: N/A
    Stream #0:0: Video: hevc (Main), yuv420p(tv), 640x480 [SAR 1:1 DAR 4:3], 1 fps, 1 tbr, 1200k tbn, 1 tbc
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> png (native))
Press [q] to stop, [?] for help
Output #0, image2, to 'frame_%05d.png':
  Metadata:
    encoder         : Lavf58.45.100
    Stream #0:0: Video: png, rgb24, 640x480 [SAR 1:1 DAR 4:3], q=2-31, 200 kb/s, 1 fps, 1 tbn, 1 tbc
    Metadata:
      encoder         : Lavc58.91.100 png
frame=   20 fps=0.0 q=-0.0 Lsize=N/A time=00:00:20.00 bitrate=N/A speed= 837x    
video:342kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
```
Perhaps the video is too short and have challenging contents. We can try to increase bitrate even more for this codec, what do you think?

Comment (alalek) [reply modules/videoio/src/cap_mfx_writer.cpp]: > encoded videos are broken

I mean in this two stage tests of encoding + decoding, the first encoding part is broken and provides incorrect results (check frame.png images).

Comment (mshabunin) [reply modules/videoio/src/cap_mfx_writer.cpp]: Aah, yes. It is usually the encoding which breaks in these tests 😃, that's why I tried to increase the bitrate. I don't see what else we can do here, maybe upgrading the driver on the test machine would help.

**Pull request merged**